pycqed/measurement/quantum_experiment.py

```diff
@@ -520,19 +532,48 @@ def _configure_mc(self, MC=None):
                 unit = list(self.sweep_points[1].values())[0][2]
             except TypeError:
                 sweep_param_name, unit = "None", ""
-            if len(self.channels_to_upload) == 0:
-                self.channels_to_upload = "all"
             if self.sweep_functions[1] == awg_swf.SegmentSoftSweep:
-                self.MC.set_sweep_function_2D(self.sweep_functions[1](
-                    sweep_func_1st_dim, self.sequences, sweep_param_name, unit,
-                    self.channels_to_upload))
+                sweep_func_2nd_dim = self.sweep_functions[1](
+                    sweep_func_1st_dim, self.sequences, sweep_param_name, unit)
             else:
                 # In case of an unknown sweep function type, it is assumed
                 # that self.sweep_functions[1] has already been initialized
                 # with all required parameters and can be directly passed to
                 # MC.
-                self.MC.set_sweep_function_2D(self.sweep_functions[1])
+                sweep_func_2nd_dim = self.sweep_functions[1]
+
+            if self.filter_segments_mask is not None and \
+                    self.compression_seg_lim is not None:
+                log.warning("Combining compression_seg_lim and "
+                            "filter_segments_mask is not supported. Ignoring "
+                            "filter_segments_mask.")
+            elif self.filter_segments_mask is not None:
+                mask = np.array(self.filter_segments_mask)
+                # Only segments with indices included in the mask can be
+                # filtered out. The others will always be measured.
+                for seq in self.sequences:
+                    for i, seg in enumerate(seq.segments.values()):
+                        if i < mask.shape[0]:
+                            seg.allow_filter = True
+                # Create filter lookup table for FilteredSweep
+                lookup = {}
+                for i, sp in enumerate(self.mc_points[1]):
+                    if i >= mask.shape[1]:
+                        # measure everything
+                        lookup[sp] = (0, 32767)
+                    elif True in mask[:, i]:
+                        # measure from the first True up to the last True
+                        lookup[sp] = (
+                            list(mask[:, i]).index(True),
+                            mask.shape[0] - list(mask[:, i])[::-1].index(
+                                True) - 1)
+                    else:
+                        # measure nothing (by setting last < first)
+                        lookup[sp] = (1, 0)
+                sweep_func_2nd_dim = swf.FilteredSweep(
+                        self.sequences[0], lookup, [sweep_func_2nd_dim])
```

Comment:
Yes, it will measure the three points in the middle, but no, it will not fail. You can think of this as a way of automatically correcting the mask to something that is supported by the lower layers. I have added further comments to the code and I have renamed the variable to the same name as in FilteredSweep.